### PR TITLE
【PIR Dist Op Reg No.28】 reg dgc

### PR DESCRIPTION
**Pull request by xiaoyewww**
### PR Category
Execute Infrastructure

### PR Types
Devs

### Description

https://github.com/PaddlePaddle/Paddle/issues/60436
注册算子dgc

### CODE REVIEW

File: paddle/fluid/pir/dialect/operator/ir/ops.yaml [hunk 1]
```diff
@@ -454,6 +454,13 @@
   optional : in_accum, in_state, out_scale, out_accum, out_state
   inplace : (scale -> out_scale, in_accum -> out_accum, in_state -> out_state)
 
+- op : dgc
+  args : (Tensor u, Tensor v, Tensor grad, Tensor param, Tensor current_step, Tensor nranks, float[] sparsity, , float m=0.9, bool use_nesterov=true, float rampup_begin_step=0.0, float rampup_step=0.0, float regular_coeff=0.0, int regular_type=0)
```
Comment (kangguangli) [inline paddle/fluid/pir/dialect/operator/ir/ops.yaml]: 多打了逗号

Comment (xiaoyewww) [reply paddle/fluid/pir/dialect/operator/ir/ops.yaml]: 感谢已修改，请问一下新的报错是注册算子的时候还有地方写错了吗

**Review by kangguangli — COMMENTED** [review]
从代码来看，https://github.com/PaddlePaddle/Paddle/blob/65126fa8feaba8a1e88a940f00707824df5a7e83/paddle/fluid/operators/CMakeLists.txt#L134-L135

这里应该是对应流水线没有开启WITH_DGC选项，需要修改CMakeLists，
在第30行前加入这样的判断：
```
if(NOT WITH_DGC)
  list(REMOVE_ITEM TEST_INTERP_CASES test_dgc_translate)
endif()
```

**Comment by xiaoyewww** [comment]
> 从代码来看，
> 
> https://github.com/PaddlePaddle/Paddle/blob/65126fa8feaba8a1e88a940f00707824df5a7e83/paddle/fluid/operators/CMakeLists.txt#L134-L135
> 
> 这里应该是对应流水线没有开启WITH_DGC选项，需要修改CMakeLists， 在第30行前加入这样的判断：
> 
> ```
> if(NOT WITH_DGC)
>   list(REMOVE_ITEM TEST_INTERP_CASES test_dgc_translate)
> endif()
> ```

多谢，已修改

File: test/ir/pir/translator/CMakeLists.txt [hunk 2]
```diff
@@ -17,7 +17,7 @@ list(APPEND DISTRIBUTED_OP_TRANSLATOR_TEST
      test_distributed_lookup_table_translate)
 list(APPEND DISTRIBUTED_OP_TRANSLATOR_TEST
      test_distributed_push_sparse_translator)
-list(APPEND DISTRIBUTED_OP_TRANSLATOR_TEST test_distributed_fused_lamb_init)
+list(APPEND DISTRIBUTED_OP_TRANSLATOR_TEST test_dgc_translator)
```
Comment (kangguangli) [inline test/ir/pir/translator/CMakeLists.txt:21]:
```suggestion
test_dgc_translator
```
这个文件存在吗？我在PR里怎么没看到

Comment (xiaoyewww) [reply test/ir/pir/translator/CMakeLists.txt:21]: 不好意思，忘记添加了

**Comment by xiaoyewww** [comment]
> 从代码来看，
> 
> https://github.com/PaddlePaddle/Paddle/blob/65126fa8feaba8a1e88a940f00707824df5a7e83/paddle/fluid/operators/CMakeLists.txt#L134-L135
> 
> 这里应该是对应流水线没有开启WITH_DGC选项，需要修改CMakeLists， 在第30行前加入这样的判断：
> 
> ```
> if(NOT WITH_DGC)
>   list(REMOVE_ITEM TEST_INTERP_CASES test_dgc_translate)
> endif()
> ```

这里这样修改后似乎还是遇到老问题

File: paddle/fluid/operators/CMakeLists.txt [hunk 3]
```diff
@@ -27,6 +27,10 @@ add_subdirectory(string)
 add_subdirectory(prim_ops)
 
 
+if(NOT WITH_DGC)
+  list(REMOVE_ITEM TEST_INTERP_CASES test_dgc_translator)
+endif()
+
```
Comment (kangguangli) [inline paddle/fluid/operators/CMakeLists.txt]: 加错地方了，加在test/ir/pir/translator/CMakeLists.txt里

Comment (xiaoyewww) [reply paddle/fluid/operators/CMakeLists.txt]: 抱歉，不知道怎么加在那边了

**Comment by xiaoyewww** [comment]
@kangguangli 这个麻烦再看一下，还是原来的问题

**Comment by kangguangli** [comment]
现在的报错是在翻译过程中报错：
> 2024-03-26 19:29:43 ======================================================================
2024-03-26 19:29:43 ERROR: test_translator (test_dgc_translator.TestDgcOpTranslator)
2024-03-26 19:29:43 ----------------------------------------------------------------------
2024-03-26 19:29:43 Traceback (most recent call last):
2024-03-26 19:29:43   File "/mnt/paddle/build/test/ir/pir/translator/test_dgc_translator.py", line 73, in test_translator
2024-03-26 19:29:43     self.check()
2024-03-26 19:29:43   File "/mnt/paddle/build/test/ir/pir/translator/test_op_translator.py", line 42, in check
2024-03-26 19:29:43     pir_program = pir.translate_to_pir(self.main_program.desc)
2024-03-26 19:29:43 ValueError: 
2024-03-26 19:29:43 
2024-03-26 19:29:43 --------------------------------------
2024-03-26 19:29:43 C++ Traceback (most recent call last):
2024-03-26 19:29:43 --------------------------------------
2024-03-26 19:29:43 0   paddle::TranslateLegacyProgramToProgram(paddle::framework::ProgramDesc const&)
2024-03-26 19:29:43 1   paddle::translator::ProgramTranslator::Translate()
2024-03-26 19:29:43 2   paddle::translator::ProgramTranslator::TranslateBlock(paddle::framework::BlockDesc const&, unsigned long, unsigned long, paddle::translator::TranslationContext*, pir::Block*)
2024-03-26 19:29:43 3   paddle::translator::ProgramTranslator::TranslateGeneralOperation(paddle::framework::OpDesc const*, paddle::translator::TranslationContext*, pir::Block*)
2024-03-26 19:29:43 4   std::_Function_handler<pir::Operation* (pir::IrContext*, paddle::translator::TranslationContext*, paddle::framework::OpDesc const&, pir::Block*), paddle::translator::OpTranscriber>::_M_invoke(std::_Any_data const&, pir::IrContext*&&, paddle::translator::TranslationContext*&&, paddle::framework::OpDesc const&, pir::Block*&&)
2024-03-26 19:29:43 5   paddle::translator::OpTranscriber::operator()(pir::IrContext*, paddle::translator::TranslationContext*, paddle::framework::OpDesc const&, pir::Block*)
2024-03-26 19:29:43 6   common::enforce::EnforceNotMet::EnforceNotMet(common::ErrorSummary const&, char const*, int)
2024-03-26 19:29:43 7   common::enforce::GetCurrentTraceBackString[abi:cxx11](bool)
2024-03-26 19:29:43 
2024-03-26 19:29:43 ----------------------
2024-03-26 19:29:43 Error Message Summary:
2024-03-26 19:29:43 ----------------------
2024-03-26 19:29:43 InvalidArgumentError: Input param not found when parsing op dgc
2024-03-26 19:29:43   [Hint: Expected legacy_input_vars.size() == 1UL, but received legacy_input_vars.size():0 != 1UL:1.] (at /paddle/paddle/fluid/ir_adaptor/translator/op_translator.cc:560)

这个错误是说翻译的时候在旧IR算子中没有发现 param 这个输入，一般是由于param为空导致的，原因是 旧IR下有些输入没有正确的声明optional，后续遇到此类问题，可以在算子的yaml里添加下面的声明试试。
```
   optional: param
```

**Comment by xiaoyewww** [comment]
> 现在的报错是在翻译过程中报错：
> 
> > 2024-03-26 19:29:43 ======================================================================
> > 2024-03-26 19:29:43 ERROR: test_translator (test_dgc_translator.TestDgcOpTranslator)
> > 2024-03-26 19:29:43 ----------------------------------------------------------------------
> > 2024-03-26 19:29:43 Traceback (most recent call last):
> > 2024-03-26 19:29:43   File "/mnt/paddle/build/test/ir/pir/translator/test_dgc_translator.py", line 73, in test_translator
> > 2024-03-26 19:29:43     self.check()
> > 2024-03-26 19:29:43   File "/mnt/paddle/build/test/ir/pir/translator/test_op_translator.py", line 42, in check
> > 2024-03-26 19:29:43     pir_program = pir.translate_to_pir(self.main_program.desc)
> > 2024-03-26 19:29:43 ValueError:
> > 2024-03-26 19:29:43
> > 2024-03-26 19:29:43 --------------------------------------
> > 2024-03-26 19:29:43 C++ Traceback (most recent call last):
> > 2024-03-26 19:29:43 --------------------------------------
> > 2024-03-26 19:29:43 0   paddle::TranslateLegacyProgramToProgram(paddle::framework::ProgramDesc const&)
> > 2024-03-26 19:29:43 1   paddle::translator::ProgramTranslator::Translate()
> > 2024-03-26 19:29:43 2   paddle::translator::ProgramTranslator::TranslateBlock(paddle::framework::BlockDesc const&, unsigned long, unsigned long, paddle::translator::TranslationContext*, pir::Block*)
> > 2024-03-26 19:29:43 3   paddle::translator::ProgramTranslator::TranslateGeneralOperation(paddle::framework::OpDesc const*, paddle::translator::TranslationContext*, pir::Block*)
> > 2024-03-26 19:29:43 4   std::_Function_handler<pir::Operation* (pir::IrContext*, paddle::translator::TranslationContext*, paddle::framework::OpDesc const&, pir::Block*), paddle::translator::OpTranscriber>::_M_invoke(std::_Any_data const&, pir::IrContext*&&, paddle::translator::TranslationContext*&&, paddle::framework::OpDesc const&, pir::Block*&&)
> > 2024-03-26 19:29:43 5   paddle::translator::OpTranscriber::operator()(pir::IrContext*, paddle::translator::TranslationContext*, paddle::framework::OpDesc const&, pir::Block*)
> > 2024-03-26 19:29:43 6   common::enforce::EnforceNotMet::EnforceNotMet(common::ErrorSummary const&, char const*, int)
> > 2024-03-26 19:29:43 7   common::enforce::GetCurrentTraceBackString[abi:cxx11](bool)
> > 2024-03-26 19:29:43
> > 2024-03-26 19:29:43 ----------------------
> > 2024-03-26 19:29:43 Error Message Summary:
> > 2024-03-26 19:29:43 ----------------------
> > 2024-03-26 19:29:43 InvalidArgumentError: Input param not found when parsing op dgc
> > 2024-03-26 19:29:43   [Hint: Expected legacy_input_vars.size() == 1UL, but received legacy_input_vars.size():0 != 1UL:1.] (at /paddle/paddle/fluid/ir_adaptor/translator/op_translator.cc:560)
> 
> 这个错误是说翻译的时候在旧IR算子中没有发现 param 这个输入，一般是由于param为空导致的，原因是 旧IR下有些输入没有正确的声明optional，后续遇到此类问题，可以在算子的yaml里添加下面的声明试试。
> 
> ```
>    optional: param
> ```

谢谢，已这样修改，但有点困惑，哪里可以看到旧IR算子中缺少这个输入呢

**Comment by kangguangli** [comment]
从报错提示中可以看出旧IR中缺少这个参数，如果对相关代码感兴趣，可以看看这个目录下的代码paddle/fluid/ir_adaptor/translator，PFCC仓库应该也有一份文档。

另外，PR模版现在有变更，麻烦按照 https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE 修改一下，我们的任务应该属于 执行架构（Execute Infrastructure）和 框架本身开发的改动。

**Comment by xiaoyewww** [comment]
> Details

多谢已修改

File: paddle/fluid/pir/dialect/operator/ir/ops.yaml [hunk 4]
```diff
@@ -454,6 +454,13 @@
   optional : in_accum, in_state, out_scale, out_accum, out_state
   inplace : (scale -> out_scale, in_accum -> out_accum, in_state -> out_state)
 
+- op : dgc
+  args : (Tensor u, Tensor v, Tensor grad, Tensor param, Tensor current_step, Tensor nranks, float[] sparsity, float m=0.9, bool use_nesterov=true, float rampup_begin_step=0.0, float rampup_step=0.0, float regular_coeff=0.0, int regular_type=0)
+  output : Tensor(u_out), Tensor(v_out), Tensor(encode_grad), Tensor(grad_out), Tensor(k), Tensor(gather_buff)
+  kernel :
+    func : dgc
```
Comment (kangguangli) [inline paddle/fluid/pir/dialect/operator/ir/ops.yaml]:
```suggestion
    func : dgc
    param: [u, v, ...]
```
在kernel里面需要指定传给dgc_kernel参数，顺序和kernel参数一致，其他地方不用修改

Comment (xiaoyewww) [reply paddle/fluid/pir/dialect/operator/ir/ops.yaml]: 多谢，已修改，请问一下这里为什么要添加呢

**Comment by xiaoyewww** [comment]
@kangguangli 请问这个pr模版有什么问题吗，按照格式修改过了？

File: paddle/fluid/pir/dialect/operator/ir/ops.yaml [hunk 5]
```diff
@@ -454,6 +454,14 @@
...
+    param : [u, v, grad, param, current_step, nranks, sparsity, m, use_nesterov, rampup_begin_step, rampup_step, regular_coeff, regular_type]
```
Comment (kangguangli) [inline paddle/fluid/pir/dialect/operator/ir/ops.yaml]: 这个顺序和kernel顺序不完全一致，再和paddle/phi/kernels/gpu/dgc_kernel.cu里的DGCKernel对照下

**Review by kangguangli — COMMENTED** [review]
PR模版需要把原来的注释部分删除

File: paddle/fluid/pir/dialect/operator/ir/ops.yaml (hunk 5 above)
Comment (xiaoyewww) [reply paddle/fluid/pir/dialect/operator/ir/ops.yaml]: 已修改，因为这里我看sparsity没有提到默认值所以提前了一下

**Comment by xiaoyewww** [comment]
@kangguangli 这个麻烦review一下，windows ci挂了

**Review by kangguangli — APPROVED** [review]

File: test/ir/pir/translator/test_dgc_translator.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline test/ir/pir/translator/test_dgc_translator.py:1]: 2023 -> 2024

**Review by zyfncg — APPROVED** [review]

**Pull request merged**